### PR TITLE
Add new reporting system

### LINKED ISSUE #162: Refactor report building so that it can be used outside of expectations.

### PR DESCRIPTION
**Pull request by jmalloc**
#### What change does this introduce?

This PR adds the `report.Report` type and several associated types which represent a human-readable report about the activity within and final result of a test.

#### What issues does this relate to?

- Fixes #162 
- Prepares for #118 
- Prepares for #161

#### Why make this change?

To allow for a consist report building and rendering experience regardless of whether the test result was determined by an `Expectation` or by some other occurrence.

#### Is there anything you are unsure about?

No

### CODE REVIEW

File: report/builder.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package report
+
+// Builder is a utility for constructing test reprots.
```
Comment (koden-km) [inline report/builder.go:3]:
```suggestion
// Builder is a utility for constructing test reports.
```

File: report/suggestion.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	// period, exclamation or question mark, similar to how Go error messages
+	// are formatted.
+	Caption string
+}
+
+// Confidence is a subjective rating of a suggestion's likely effectiveness.
+type Confidence uint8
+
+// AbsoluteConfidence indicates that a suggestion is a definitive and
+// appropriate solution the problem described by a negative finding.
```
Comment (koden-km) [inline report/suggestion.go:30]:
```suggestion
// appropriate solution to the problem described by a negative finding.
```

File: report/builder.go (hunk 1 above)
Comment (jmalloc) [reply report/builder.go]: This type has been removed, at least for now.

File: report/step.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	Caption string
+
+	// Transcript is the set of facts emitted by the engine.
+	Transcript []TranscriptEntry
+
+	// Findings is the set of discoveries made by analysing the transcript.
+	Findings []Finding
+}
+
+// TranscriptEntry is an entry within an XXX's transcript.
```
Comment (koden-km) [inline report/step.go:24]:
```suggestion
// TranscriptEntry is an entry within a Step's transcript.
```

**Pull request closed without merging**